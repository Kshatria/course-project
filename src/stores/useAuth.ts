import { create } from 'zustand';

type AuthState = {
  isAuth: boolean;
  login: (token: string) => void;
  logout: () => void;
  token: null | string;
};

const useAuthStore = create<AuthState>((set) => ({
  token: null,
  isAuth: !!localStorage.getItem('token'),
  login: (token) => {
    localStorage.setItem('token', token);
    set({ token, isAuth: true });
  },
  logout: () => {
    localStorage.removeItem('token');
    set({ token: null, isAuth: false });
  },
}));

const useAuth = () => {
  const { isAuth, login, logout } = useAuthStore();
  return { isAuth, login, logout };
};

export { useAuth };
