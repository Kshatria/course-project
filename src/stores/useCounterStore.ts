import { create } from 'zustand';
import { immer } from 'zustand/middleware/immer';

type CounterState = {
  count: number;
  increment: () => void;
  decrement: () => void;
  reset: () => void;
};

export const useCounterStore = create<CounterState>()(
  immer((set) => ({
    count: 0,
    increment: () =>
      set((state) => {
        state.count += 1;
      }),
    decrement: () =>
      set((state) => {
        state.count -= 1;
      }),
    reset: () => set({ count: 0 }),
  })),
);
