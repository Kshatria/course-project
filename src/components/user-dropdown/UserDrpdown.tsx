import { useEffect, useRef, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { FaUser } from 'react-icons/fa';
import { useAuth } from '@/stores/useAuth';
import styles from './UserDropdown.module.css'; // Импорт стилей

const UserDropdown = () => {
  const [isOpen, setIsOpen] = useState(false);
  const dropdownRef = useRef<HTMLDivElement>(null);
  const navigate = useNavigate();
  const { logout } = useAuth();

  useEffect(() => {
    const handleClickOutside = (e: MouseEvent) => {
      if (dropdownRef.current && !dropdownRef.current.contains(e.target as Node)) {
        setIsOpen(false);
      }
    };

    document.addEventListener('mousedown', handleClickOutside);
    return () => document.removeEventListener('mousedown', handleClickOutside);
  }, []);

  return (
    <div className={styles.dropdown} ref={dropdownRef}>
      <button className={styles.iconButton} type="button" onClick={() => setIsOpen(!isOpen)}>
        <FaUser className="text-gray-700" />
      </button>

      {isOpen && (
        <div className={styles.menu}>
          <button
            className={styles.menuItem}
            type="button"
            onClick={() => {
              navigate('/profile');
              setIsOpen(false);
            }}
          >
            Личный кабинет
          </button>
          <button
            className={styles.menuItem}
            type="button"
            onClick={() => {
              logout();
              setIsOpen(false);
            }}
          >
            Выход
          </button>
        </div>
      )}
    </div>
  );
};

export { UserDropdown };
