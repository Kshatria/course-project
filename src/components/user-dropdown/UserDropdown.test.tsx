import { useNavigate } from 'react-router-dom';
import { fireEvent, render, screen } from '@testing-library/react';
import { useAuth } from '@/stores/useAuth';
import { UserDropdown } from './UserDrpdown';

// Мокаем зависимости
jest.mock('react-router-dom', () => ({
  useNavigate: jest.fn(),
}));

jest.mock('@/stores/useAuth', () => ({
  useAuth: jest.fn(),
}));

jest.mock('react-icons/fa', () => ({
  FaUser: () => <span data-testid="user-icon">UserIcon</span>,
}));

describe('UserDropdown', () => {
  const mockNavigate = jest.fn();
  const mockLogout = jest.fn();

  beforeEach(() => {
    (useNavigate as jest.Mock).mockReturnValue(mockNavigate);
    (useAuth as jest.Mock).mockReturnValue({
      logout: mockLogout,
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders user icon button', () => {
    render(<UserDropdown />);
    expect(screen.getByTestId('user-icon')).toBeInTheDocument();
  });

  it('opens dropdown when clicked', () => {
    render(<UserDropdown />);
    fireEvent.click(screen.getByTestId('user-icon'));
    expect(screen.getByText('Личный кабинет')).toBeInTheDocument();
    expect(screen.getByText('Выход')).toBeInTheDocument();
  });

  it('closes dropdown when clicked outside', () => {
    render(
      <div>
        <div data-testid="outside-element">Outside</div>
        <UserDropdown />
      </div>,
    );

    // Открываем dropdown
    fireEvent.click(screen.getByTestId('user-icon'));
    expect(screen.getByText('Личный кабинет')).toBeInTheDocument();

    // Кликаем снаружи
    fireEvent.mouseDown(screen.getByTestId('outside-element'));
    expect(screen.queryByText('Личный кабинет')).not.toBeInTheDocument();
  });

  it('navigates to profile when "Личный кабинет" clicked', () => {
    render(<UserDropdown />);

    // Открываем и кликаем
    fireEvent.click(screen.getByTestId('user-icon'));
    fireEvent.click(screen.getByText('Личный кабинет'));

    expect(mockNavigate).toHaveBeenCalledWith('/profile');
    expect(screen.queryByText('Личный кабинет')).not.toBeInTheDocument();
  });

  it('calls logout when "Выход" clicked', () => {
    render(<UserDropdown />);

    // Открываем и кликаем
    fireEvent.click(screen.getByTestId('user-icon'));
    fireEvent.click(screen.getByText('Выход'));

    expect(mockLogout).toHaveBeenCalled();
    expect(screen.queryByText('Выход')).not.toBeInTheDocument();
  });

  it('matches snapshot when closed', () => {
    const { asFragment } = render(<UserDropdown />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('matches snapshot when opened', () => {
    const { asFragment } = render(<UserDropdown />);
    fireEvent.click(screen.getByTestId('user-icon'));
    expect(asFragment()).toMatchSnapshot();
  });
});
