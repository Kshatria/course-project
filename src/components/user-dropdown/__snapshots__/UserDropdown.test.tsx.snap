// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserDropdown matches snapshot when closed 1`] = `
<DocumentFragment>
  <div
    class="dropdown"
  >
    <button
      class="iconButton"
      type="button"
    >
      <span
        data-testid="user-icon"
      >
        UserIcon
      </span>
    </button>
  </div>
</DocumentFragment>
`;

exports[`UserDropdown matches snapshot when opened 1`] = `
<DocumentFragment>
  <div
    class="dropdown"
  >
    <button
      class="iconButton"
      type="button"
    >
      <span
        data-testid="user-icon"
      >
        UserIcon
      </span>
    </button>
    <div
      class="menu"
    >
      <button
        class="menuItem"
        type="button"
      >
        Личный кабинет
      </button>
      <button
        class="menuItem"
        type="button"
      >
        Выход
      </button>
    </div>
  </div>
</DocumentFragment>
`;
